mm2src/mm2_p2p/src/lib.rs

```diff
@@ -0,0 +1,187 @@
+#![feature(ip)]
+
+pub mod behaviours;
+
+mod network;
+mod relay_address;
+mod swarm_runtime;
+
+use lazy_static::lazy_static;
+use secp256k1::{Message as SecpMessage, PublicKey as Secp256k1Pubkey, Secp256k1, SecretKey, SignOnly, Signature,
+                VerifyOnly};
+use serde::{de, Deserialize, Serialize, Serializer};
+use sha2::{Digest, Sha256};
+
+pub use crate::swarm_runtime::SwarmRuntime;
+
+// atomicdex related re-exports
+pub use behaviours::atomicdex::{get_gossip_mesh, get_gossip_peer_topics, get_gossip_topic_peers, get_peers_info,
+                                get_relay_mesh, spawn_gossipsub, AdexBehaviourCmd, AdexBehaviourError,
+                                AdexBehaviourEvent, AdexCmdTx, AdexEventRx, AdexResponse, AdexResponseChannel,
+                                GossipsubEvent, GossipsubMessage, MessageId, NodeType, TopicHash, WssCerts};
+
+// peers-exchange re-exports
+pub use behaviours::peers_exchange::PeerAddresses;
+
+// request-response related re-exports
+pub use behaviours::request_response::RequestResponseBehaviourEvent;
+
+// libp2p related re-exports
+pub use libp2p::identity::DecodingError;
+pub use libp2p::identity::{secp256k1::PublicKey as Libp2pSecpPublic, PublicKey as Libp2pPublic};
+pub use libp2p::{Multiaddr, PeerId};
+
+// relay-address related re-exports
+pub use network::SeedNodeInfo;
+pub use relay_address::RelayAddress;
+pub use relay_address::RelayAddressError;
+
+lazy_static! {
+    static ref SECP_VERIFY: Secp256k1<VerifyOnly> = Secp256k1::verification_only();
+    static ref SECP_SIGN: Secp256k1<SignOnly> = Secp256k1::signing_only();
+}
+
+#[derive(Clone, Copy, Debug)]
+pub enum NetworkInfo {
+    /// The in-memory network.
+    InMemory,
+    /// The distributed network (out of the app memory).
+    Distributed { network_ports: NetworkPorts },
+}
+
+impl NetworkInfo {
+    pub fn in_memory(&self) -> bool { matches!(self, NetworkInfo::InMemory) }
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct NetworkPorts {
+    pub tcp: u16,
+    pub wss: u16,
+}
+
+pub fn encode_message<T: Serialize>(message: &T) -> Result<Vec<u8>, rmp_serde::encode::Error> {
+    rmp_serde::to_vec(message)
+}
+
+#[inline]
+pub fn decode_message<'de, T: de::Deserialize<'de>>(bytes: &'de [u8]) -> Result<T, rmp_serde::decode::Error> {
+    rmp_serde::from_slice(bytes)
+}
+
+#[derive(Deserialize, Serialize)]
+struct SignedMessageSerdeHelper<'a> {
+    pubkey: PublicKey,
+    #[serde(with = "serde_bytes")]
+    signature: &'a [u8],
+    #[serde(with = "serde_bytes")]
+    payload: &'a [u8],
+}
+
+pub fn encode_and_sign<T: Serialize>(message: &T, secret: &[u8; 32]) -> Result<Vec<u8>, rmp_serde::encode::Error> {
+    let secret = SecretKey::from_slice(secret).unwrap();
```

Comment:
Are we confident we will never use this function with user defined secret keys? This `.unwrap()` can be triggered with a 0'd key or key greater than the curve order.

Reply:
I can add error handling just in case. Will work on a fix right away.

Reply:
should be fixed here https://github.com/KomodoPlatform/komodo-defi-framework/pull/2038